### DISCUSSION TITLE
How to capture an UFFTYPER message generated by embedding

### DISCUSSION THREAD

**Discussion by Jesperkemist** [rdkit/rdkit#5435] [Q&A]
Hello!

When I embed molecules
embed_sucess = AllChem.EmbedMolecule(mol, useExpTorsionAnglePrefs=True, useBasicKnowledge=True, useRandomCoords=True)
I sometimes get the output similar to the one below 
[08:48:09] UFFTYPER: Unrecognized atom type: Zn1+2 (17)

It does not seem to be an error per se as embed_sucess = 0, but when I run the simple geometry optimization routine with molecules that generate those kinds of messages, 
conformed = AllChem.UFFOptimizeMolecule(mol, maxIters=1000)
they tend to give geometries that not are reasonable (even if conformed = 0) 

I would thus like to discard those molecules in my further analysis. To do that I need to capture the UFFTYPER message in some way and this is where I fail. It is printed on the screen, but it is not returned by the AllChem.EmbedMolecule() function, it do not seem to end up in the sys.stdout, and it does not seems to generate an exception.

Here I am at a loss. If I loop over a few thousand molecules, how do I identify all that generate an UFFTYPER message? 

Best regards!
/Jesper Jacobsson

**Comment by ptosco** [comment]
@Jesperkemist RDKit 2022.03 makes capturing C++ logs much easier than it used to be.
See below for how to do that by either redirecting logs to `stderr` and capturing the `stderr` stream, or by redirecting logs to the Python logger and capturing that.

## Redirect to `stderr` and capture `stderr`
```
from contextlib import redirect_stderr
from io import StringIO
import rdkit
from rdkit import rdBase, Chem
from rdkit.Chem import AllChem

# redirect RDKit logs to stderr
rdBase.LogToPythonStderr()

mol = Chem.MolFromSmiles("[Zn+2]")

with StringIO() as buf:
    with redirect_stderr(buf):
        embed_success = AllChem.EmbedMolecule(mol, useExpTorsionAnglePrefs=True, useBasicKnowledge=True, useRandomCoords=True)
        res = buf.getvalue()

res
'[10:54:38] UFFTYPER: Unrecognized charge state for atom: 0\n[10:54:38] UFFTYPER: Unrecognized atom type: Zn+2 (0)\n'
```

## Redirect to Python's logger and capture that

```
import logging
import os
import rdkit
from rdkit import rdBase, Chem
from rdkit.Chem import AllChem

# redirect logs to Python logger
rdBase.LogToPythonLogger()

class CaptureLogger(logging.Handler):
    """Helper class that captures Python logger output"""

    def __init__(self, module=None):
        super(CaptureLogger, self).__init__(level=logging.DEBUG)
        self.logs = {}
        self.devnull = open(os.devnull, 'w')
        rdkit.log_handler.setStream(self.devnull)
        rdkit.logger.addHandler(self)

    def __enter__(self):
        return self.logs

    def __exit__(self, *args):
        self.release()

    def handle(self, record):
        key = record.levelname
        val = self.format(record)
        self.logs[key] = self.logs.get(key, "") + val
        return False

    def release(self):
        rdkit.log_handler.setStream(sys.stderr)
        rdkit.logger.removeHandler(self)
        self.devnull.close()
        return self.logs

mol = Chem.MolFromSmiles("[Zn+2]")

with CaptureLogger() as capture:
    embed_success = AllChem.EmbedMolecule(mol, useExpTorsionAnglePrefs=True, useBasicKnowledge=True, useRandomCoords=True)

capture["ERROR"]
'[11:00:30] UFFTYPER: Unrecognized charge state for atom: 0[11:00:30] UFFTYPER: Unrecognized atom type: Zn+2 (0)'
```
